core/rawdb/prunedfreezer.go

```diff
@@ -0,0 +1,286 @@
+			case <-f.quit:
+				return
+			}
+		}
+
+		// Retrieve the freezing threshold.
+		hash := ReadHeadBlockHash(nfdb)
+		if hash == (common.Hash{}) {
+			log.Debug("Current full block hash unavailable") // new chain, empty database
+			backoff = true
+			continue
+		}
+		number := ReadHeaderNumber(nfdb, hash)
+		threshold := atomic.LoadUint64(&f.threshold)
+
+		switch {
+		case number == nil:
+			log.Error("Current full block number unavailable", "hash", hash)
+			backoff = true
+			continue
+
+		case *number < threshold:
+			log.Debug("Current full block not old enough", "number", *number, "hash", hash, "delay", threshold)
+			backoff = true
+			continue
+
+		case *number-threshold <= f.frozen:
+			log.Debug("Ancient blocks frozen already", "number", *number, "hash", hash, "frozen", f.frozen)
+			backoff = true
+			continue
+		}
+		head := ReadHeader(nfdb, hash, *number)
+		if head == nil {
+			log.Error("Stable state block unavailable", "number", *number, "hash", hash)
+			backoff = true
+			continue
+		}
+
+		stableStabeNumber := ReadStableStateBlockNumber(nfdb)
+		switch {
+		case stableStabeNumber < params.StableStateThreshold:
+			log.Debug("Stable state block not old enough", "number", stableStabeNumber)
+			backoff = true
+			continue
+
+		case stableStabeNumber > *number:
+			log.Warn("Stable state block biger current full block", "number", stableStabeNumber, "number", *number)
+			backoff = true
+			continue
+		}
+		stableStabeNumber -= params.StableStateThreshold
+
+		// Seems we have data ready to be frozen, process in usable batches
+		limit := *number - threshold
+		if limit > stableStabeNumber {
+			limit = stableStabeNumber
+		}
+
+		if limit < f.frozen {
+			log.Debug("Stable state block has prune", "limit", limit, "frozen", f.frozen)
+			backoff = true
+			continue
+		}
+
+		if limit-f.frozen > freezerBatchLimit {
+			limit = f.frozen + freezerBatchLimit
+		}
+		var (
+			start    = time.Now()
+			first    = f.frozen
+			ancients = make([]common.Hash, 0, limit-f.frozen)
+		)
+		for f.frozen <= limit {
+			// Retrieves all the components of the canonical block
+			hash := ReadCanonicalHash(nfdb, f.frozen)
+			if hash == (common.Hash{}) {
+				log.Error("Canonical hash missing, can't freeze", "number", f.frozen)
+			}
+			log.Trace("Deep froze ancient block", "number", f.frozen, "hash", hash)
+			// Inject all the components into the relevant data tables
+			if err := f.AppendAncient(f.frozen, nil, nil, nil, nil, nil); err != nil {
+				log.Error("Append ancient err", "number", f.frozen, "hash", hash, "err", err)
+				break
+			}
+			if hash != (common.Hash{}) {
+				ancients = append(ancients, hash)
+			}
+		}
+		// Batch of blocks have been frozen, flush them before wiping from leveldb
+		if err := f.Sync(); err != nil {
+			log.Crit("Failed to flush frozen tables", "err", err)
+		}
+		// Batch of blocks have been frozen, flush them before wiping from leveldb
+		backoff = f.frozen-first >= freezerBatchLimit
```

Comment:
In most cases, `backoff` is false, will this keep this routine too busy to occupy one core of CPU, which is not necessary?